File: coin/Eonix/src/blockchain.py
# blockchain.py
import hashlib
import time

class Block:
    def __init__(self, index, previous_hash, timestamp, transactions, nonce=0):
        self.index = index
        self.previous_hash = previous_hash
        self.timestamp = timestamp
        self.transactions = transactions
        self.nonce = nonce
        self.hash = self.calculate_hash()

    def calculate_hash(self):
        data = str(self.index) + self.previous_hash + str(self.timestamp) + str(self.transactions) + str(self.nonce)
        return hashlib.sha256(data.encode()).hexdigest()

class Blockchain:
    def __init__(self):
        self.chain = [self.create_genesis_block()]
        self.difficulty = 2
        self.pending_transactions = []

    def create_genesis_block(self):
        return Block(0, "0", int(time.time()), [])

    def get_latest_block(self):
        return self.chain[-1]

    def add_block(self, new_block):
        new_block.previous_hash = self.get_latest_block().hash
        new_block.hash = new_block.calculate_hash()
        self.chain.append(new_block)

    def add_transaction(self, transaction):
        self.pending_transactions.append(transaction)

    def mine_pending_transactions(self):
        if not self.pending_transactions:
            return False

        new_block = Block(len(self.chain), self.get_latest_block().hash, int(time.time()), self.pending_transactions)
        new_block.mine_block(self.difficulty)
        self.chain.append(new_block)
        self.pending_transactions = []
        return True

    def validate_chain(self):
        for i in range(1, len(self.chain)):
            current_block = self.chain[i]
            previous_block = self.chain[i - 1]
            if current_block.hash != current_block.calculate_hash():
                return False
            if current_block.previous_hash != previous_block.hash:
                return False
        return True

File: blockchain_integration/pi_network/AegisPi/src/blockchain/blockchain.js
import { BlockchainUtils } from './blockchain_utils';
import { Crypto } from './crypto';

class Block {
  constructor(index, previousHash, timestamp, data, hash) {
    this.index = index;
    this.previousHash = previousHash;
    this.timestamp = timestamp;
    this.data = data;
    this.hash = hash;
  }
}

class Blockchain {
  constructor() {
    this.chain = [this.createGenesisBlock()];
    this.pendingTransactions = [];
    this.miningDifficulty = 3;
    this.blockSizeLimit = 10;
  }

  createGenesisBlock() {
    return new Block(0, '0', 0, 'Genesis Block', '0');
  }

  getLatestBlock() {
    return this.chain[this.chain.length - 1];
  }

  addTransaction(transaction) {
    this.pendingTransactions.push(transaction);
  }

  minePendingTransactions(minerAddress) {
    const block = this.createBlock(this.pendingTransactions, minerAddress);
    this.chain.push(block);
    this.pendingTransactions = [];
    return block;
  }

  createBlock(transactions, minerAddress) {
    const previousBlock = this.getLatestBlock();
    const index = previousBlock.index + 1;
    const timestamp = Date.now();
    const data = transactions;
    const hash = BlockchainUtils.calculateHash(index, previousBlock.hash, timestamp, data);
    return new Block(index, previousBlock.hash, timestamp, data, hash);
  }

  validateChain() {
    for (let i = 1; i < this.chain.length; i++) {
      const currentBlock = this.chain[i];
      const previousBlock = this.chain[i - 1];
      if (currentBlock.hash !== BlockchainUtils.calculateHash(currentBlock.index, previousBlock.hash, currentBlock.timestamp, currentBlock.data)) {
        return false;
      }
    }
    return true;
  }

  static async getBlockchainInstance() {
    const blockchain = new Blockchain();
    await blockchain.initialize();
    return blockchain;
  }

  async initialize() {
    // Initialize the blockchain by loading the chain from storage or creating a new one
  }
}

export { Blockchain, Block };
